### ISSUE TITLE
`list[str]` is not valid input to `list[str | int]` in some cases

### ISSUE THREAD

**Issue by samuelcolvin** [microsoft/pyright#4313]
See the code below. 

In short, `[['hello']]` is not considered a valid input to `Union[str, List[Union[str, int]], List[List[Union[str, int]]]]` which it should be - it matches the last case.

If I simply the annotation down to `List[List[Union[str, int]]]`, and pass the same argument, pyright accepts it.

```py
from typing import Union, List

def foobar(thing: Union[str, List[Union[str, int]], List[List[Union[str, int]]]]) -> None:
    print(thing)


# all ok:
foobar('hello')
foobar(['hello', 1])
foobar([['hello', 1], ['world', 2]])
# wrongly causes error:
foobar([['world']])
```

Output:
```
No configuration file found.
pyproject.toml file found at /Users/samuel/code/pydantic-core.
Loading pyproject.toml file at /Users/samuel/code/pydantic-core/pyproject.toml
Assuming Python version 3.10
Assuming Python platform Darwin
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath /Users/samuel/code/pydantic-core/typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.283
/Users/samuel/code/pydantic-core/pyright-demo.py
  /Users/samuel/code/pydantic-core/pyright-demo.py:12:9 - error: Argument of type "list[list[str]]" cannot be assigned to parameter "thing" of type "str | List[str | int] | List[List[str | int]]" in function "foobar"
    Type "list[str]" cannot be assigned to type "str | int"
      "list[str]" is incompatible with "str"
      "list[str]" is incompatible with "int" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations 
Completed in 0.439sec
(env310) ~/code/pydantic-core 1 ➤ pyright --version
pyright 1.1.283
(env310) ~/code/pydantic-core 0 ➤ python -V
Python 3.10.4
```

**Comment by samuelcolvin** [comment]
Interestingly, if I change the order of the union members to `Union[str, List[List[Union[str, int]]], List[Union[str, int]]]`, it passes.

**Comment by erictraut** [comment]
Yes, this is expected behavior.

It's a side effect of the way pyright performs bidirectional type inference when the "expected type" is a union. In your case, the expected type is `str | list[str | int] | list[list[str | int]]`. Bidirectional type inference is performed on a specific (non-union) subtype, so pyright determines which subtype is potentially a match for a list. In most cases, there is a single `list` or `Sequence` subtype, and it uses that for bidirectional type inference. In your example, there are two candidates, both of them `list`. It therefore chooses the first one and attempts to use it for bidirectional type inference. 

Addressing this limitation is theoretically possible, but it would add significant complexity and result in a big performance hit within the type analyzer. Since this is such a rare edge case (this is the first time I've seen it reported in 3 years), I don't think it's worth addressing. You can work around it in your code by using an unambiguous type declaration.

```python
temp: list[list[str | int]] = [["world"]]
foobar(temp)
```

**Comment by samuelcolvin** [comment]
Ok, thanks so much for the explanation.

**Comment by erictraut** [comment]
I just thought of a way that I might be able to implement this without a significant performance hit to the common case. I'm going to re-open the issue and investigate the feasibility of this approach.

**Comment by erictraut** [comment]
My idea worked. This will be addressed in the next release.

**Comment by samuelcolvin** [comment]
Amazing, thank you.

**Comment by erictraut** [comment]
This is included in pyright 1.1.284, which I just published. It will also be included in a future release of pylance.

**Comment by swats-the-floran** [comment]
python 1.1.356 got the same problem.

**Comment by erictraut** [comment]
@swats-the-floran, this issue was fixed and closed years ago. If you think that you have discovered a new bug, please open a new issue. Also, when reporting issues, please include a self-contained, minimal code sample that demonstrates the problem. Your code sample above relies on symbols that are presumably imported from some library (`BaseModel`, `Field`, `field_validator`), but you don't specify which library (or version) or how you're importing these symbols.
